### ISSUE TITLE
Add support for missing values in the encoders

### ISSUE THREAD

**Issue by GaelVaroquaux** [dirty-cat/dirty_cat#124]
Encoding a missing value as a vector of zeros is a reasonable thing. Our theoretical study (https://arxiv.org/abs/1902.06931) shows that the most important thing is to encode them in a special value that can be later picked up by the supervised step.

Our encoders should have an option that controls whether missing values are encoded as zeros or an error is raised (following scikit-learn encoders).

**Comment by amy12xx** [comment]
Is this like the parameter handle_unknown{‘error’, ‘ignore’}, default=’error’ in sklearn's one hot encoder?

**Comment by GaelVaroquaux** [comment]
Thanks for asking!

Not quite: handle_unknown governs everything that happens for categories not seen in the train set. However, the encoders that we have in dirty-cat can deal with those, as long as they are strings.

Rather, it would be to deal with NaNs, which encode missingness.

**Comment by amy12xx** [comment]
Can I take a look at it? I maybe a little slow on fixing it though. 

I have yet to read your study, but is it adequate to encode all missing values with the same encoding (0) and assume they will be handled downstream or better to encode them as unique categories?

I looked briefly at the code, and the sklearn function _check_X() throws the error when there are NaNs. Would a reasonable solution be to impute the NaNs with a special value (for e.g. __nan__) before this check, and then handle later on to encode to 0?

**Comment by GaelVaroquaux** [comment]
It would be great if you'd work on this!!!

I indeed think that the simplest way to embed missing values would be to cast them to a vector of zeros.

Now, with regards to how to implement this, we are hitting a few problems.

I had a quick look, and it seems that, for the SimilarityEncoder at least, replacing nans by the empty string '' works (though it spits out nasty warnings that we should capture at the right place, ie as deep as possible).

```
import pandas as pd
import numpy as np

a = [['a',    'b'],
     ['b',    'a'],
     ['b',    np.nan],
     ['a',    'c'],
     [np.nan, 'a'],
    ]

from dirty_cat import SimilarityEncoder
se = SimilarityEncoder()

X = se.fit_transform(pd.DataFrame(a).fillna(value=''))
```

We could start with this strategy, and do a first PR for the similarity? Note that I don't think that the solution that I coded (using pd's fillna) should be implemented on all types of input. We might have to have a different treatment if the input is a list, and yet another if it is a numpy array.

Also, if this strategy ends up working for other of our encoders (eg the MinHashEncoder) we should probably introduce a subclass of _BaseEncoder that does this transformation in the _check_X method.

**Comment by amy12xx** [comment]
So this is the code I added to the fit and transform methods. I am coercing the array to numpy dtype=object if it is a list or has a dtype like unicode/string, since i did not find an optimal way to check for NaNs. If it is unicode/string, the NaNs are anyway converted to string, which does not help. 

`        if self.handle_missing not in ['error', '']:
            template = ("handle_missing should be either 'error' or "
                        "'', got %s")
            raise ValueError(template % self.handle_missing)
        if hasattr(X, 'iloc') and X.isna().values.any():
            if self.handle_missing == 'error':
                msg = ("Found missing values in input data.")
                raise ValueError(msg)
            if self.handle_missing != 'error':
                X = X.fillna(self.handle_missing)
        elif not hasattr(X, 'dtype') and isinstance(X, list):
            X = np.asarray(X, dtype=object)

        if hasattr(X, 'dtype'):
            if np.issubdtype(X.dtype, np.str_):
                X = np.asarray(X, dtype=object)
            mask = _object_dtype_isnan(X)
            if X.dtype.kind == 'O' and mask.any():
                if self.handle_missing == 'error':
                    msg = ("Found missing values in input data.")
                    raise ValueError(msg)
                if self.handle_missing != 'error':
                    X = np.ma.array(X, mask=mask)
                    X = np.ma.filled(X, self.handle_missing)`

I got these warnings when I exected:
RuntimeWarning: invalid value encountered in true_divide
in the function _ngram_similarity_one_sample_inplace(), which is executed in parallel over the data. I changed the np.divide function to handle this: https://stackoverflow.com/a/37977222

**Comment by amy12xx** [comment]
`        if self.handle_missing not in ['error', '']:
            template = ("handle_missing should be either 'error' or "
                        "'', got %s")
            raise ValueError(template % self.handle_missing)
        if hasattr(X, 'iloc') and X.isna().values.any():
            if self.handle_missing == 'error':
                msg = ("Found missing values in input data.")
                raise ValueError(msg)
            if self.handle_missing != 'error':
                X = X.fillna(self.handle_missing)
        elif not hasattr(X, 'dtype') and isinstance(X, list):
            X = np.asarray(X, dtype=object)

        if hasattr(X, 'dtype'):
            if np.issubdtype(X.dtype, np.str_):
                X = np.asarray(X, dtype=object)
            mask = _object_dtype_isnan(X)
            if X.dtype.kind == 'O' and mask.any():
                if self.handle_missing == 'error':
                    msg = ("Found missing values in input data.")
                    raise ValueError(msg)
                if self.handle_missing != 'error':
                    X = np.ma.array(X, mask=mask)
                    X = np.ma.filled(X, self.handle_missing)`

**Comment by amy12xx** [comment]
`
        if self.handle_missing not in ['error', '']:
            template = ("handle_missing should be either 'error' or "
                        "'', got %s")
            raise ValueError(template % self.handle_missing)
        if hasattr(X, 'iloc') and X.isna().values.any():
            if self.handle_missing == 'error':
                msg = ("Found missing values in input data.")
                raise ValueError(msg)
            if self.handle_missing != 'error':
                X = X.fillna(self.handle_missing)
        elif not hasattr(X, 'dtype') and isinstance(X, list):
            X = np.asarray(X, dtype=object)

`

**Comment by amy12xx** [comment]
`
        if hasattr(X, 'dtype'):
            if np.issubdtype(X.dtype, np.str_):
                X = np.asarray(X, dtype=object)
            mask = _object_dtype_isnan(X)
            if X.dtype.kind == 'O' and mask.any():
                if self.handle_missing == 'error':
                    msg = ("Found missing values in input data.")
                    raise ValueError(msg)
                if self.handle_missing != 'error':
                    X = np.ma.array(X, mask=mask)
                    X = np.ma.filled(X, self.handle_missing)

`

**Comment by amy12xx** [comment]
```
        if self.handle_missing not in ['error', '']:
            template = ("handle_missing should be either 'error' or "
                        "'', got %s")
            raise ValueError(template % self.handle_missing)
        if hasattr(X, 'iloc') and X.isna().values.any():
            if self.handle_missing == 'error':
                msg = ("Found missing values in input data.")
                raise ValueError(msg)
            if self.handle_missing != 'error':
                X = X.fillna(self.handle_missing)
        elif not hasattr(X, 'dtype') and isinstance(X, list):
            X = np.asarray(X, dtype=object)

        if hasattr(X, 'dtype'):
            if np.issubdtype(X.dtype, np.str_):
                X = np.asarray(X, dtype=object)
            mask = _object_dtype_isnan(X)
            if X.dtype.kind == 'O' and mask.any():
                if self.handle_missing == 'error':
                    msg = ("Found missing values in input data.")
                    raise ValueError(msg)
                if self.handle_missing != 'error':
                    X = np.ma.array(X, mask=mask)
                    X = np.ma.filled(X, self.handle_missing)
```

**Comment by GaelVaroquaux** [comment]
I'll have a look ASAP. One thing to note: if the dtype is string or unicode, there cannot be nans. So you don't need to coerce, you can just move on without worrying about nans/missing values. This will be the most efficient approach.

**Comment by GaelVaroquaux** [comment]
Can you send a pull request (make in "work in progress" if it's not finished). It will be easier to comment on it.

Rather than using a masked array, wouldn't something it be faster to do something like:
```
X[mask] = self.handle_missing
```

**Comment by amy12xx** [comment]
Hi @GaelVaroquaux , I checked the implementation of MinHashEncoder, here's what I found:

Line 186 checks if the first element in X is not a string. I think this could be extended for any element in X.
assert isinstance(X[0], str), "ERROR: Input data is not string."

Imputing missing values with a blank string '' does not work, since it is later computing the ASCII/UTF-8 value for the char, and throws an error then. 

It seems like just improving the assertion above could be the right fix?

**Comment by GaelVaroquaux** [comment]
The assertion could certainly be improved. Firstly, if the input is an array and has a dtype of kind 'U' or 'S' (check .dtype.kind), we now that its fine: it's an array and cannot have nans.

Then, to handle nans, I would suggest to modify the get_hash method and handle nans. The challenge is that this method is called in a very tight loop and hence should be as fast as possible. I think that see ways of refactoring this in a better way. Are you up for this? In which case we should schedule a call.

**Comment by amy12xx** [comment]
Yes, I can give it a shot. We can connect this weekend, if that works for you.

**Comment by GaelVaroquaux** [comment]
Sounds good for this week end. I'm actually packed with work until Saturday. Thanks a lot!

You'll tell me what time suits you. I was planning to go hiking on Saturday morning (my time), but I can probably squeeze something in before I leave.

Thanks heaps!

**Comment by amy12xx** [comment]
We can do Sunday, if Saturday does not work for you. I'm available both days, 10:30 AM - 9:30 PM (IST). I am not sure what time zone you are in, so go ahead and set one up in that window, and also what you suggest we can use to connect. I am on Skype/Google etc.  I'm also on the scikit gitter channel.

**Comment by amy12xx** [comment]
Fixed for existing encoders.

**Comment by GaelVaroquaux** [comment]
Thanks heaps @amy12xx  !!
